src/index.ts

```diff
@@ -184,6 +186,185 @@ export function HTPipeAttachData<
+>;
+
+// left has preAuthorize and right does not
+export function HTPipePreAuthorize<
+  TLeft extends HasPreAuthorize<
+    any,
+    | boolean
+    | (TRight extends HasPreAuthorize<any, any>
+        ? Pick<
+            Parameters<TRight['preAuthorize']>[0],
+            keyof ReturnType<
+              TLeft extends HasPreAuthorize<any, any>
+                ? TLeft['preAuthorize']
+                : () => {}
+            >
+          >
+        : any)
+  >,
+  TRight extends OptionallyHasPreAuthorize<any, any>,
+  TContextInLeft extends Parameters<TLeft['preAuthorize']>[0],
+  TContextOutLeft extends ReturnType<TLeft['preAuthorize']>
+>(left: TLeft, right: TRight): HasPreAuthorize<TContextInLeft, TContextOutLeft>;
+
+// right has preAuthorize and left does not
+export function HTPipePreAuthorize<
+  TLeft extends OptionallyHasPreAuthorize<
+    any,
+    | boolean
+    | (TRight extends HasPreAuthorize<any, any>
+        ? Pick<
+            Parameters<TRight['preAuthorize']>[0],
+            keyof ReturnType<
+              TLeft extends HasPreAuthorize<any, any>
+                ? TLeft['preAuthorize']
+                : () => {}
+            >
+          >
+        : any)
+  >,
+  TRight extends HasPreAuthorize<any, any>,
+  TContextInRight extends Parameters<TRight['preAuthorize']>[0],
+  TContextOutRight extends ReturnType<TRight['preAuthorize']>
+>(
+  left: TLeft,
+  right: TRight
+): HasPreAuthorize<TContextInRight, TContextOutRight>;
+
+// right and left doesn't have preAuthorize
+export function HTPipePreAuthorize<
+  TLeft extends OptionallyHasPreAuthorize<
+    any,
+    | boolean
+    | (TRight extends HasPreAuthorize<any, any>
+        ? Pick<
+            Parameters<TRight['preAuthorize']>[0],
+            keyof ReturnType<
+              TLeft extends HasPreAuthorize<any, any>
+                ? TLeft['preAuthorize']
+                : () => {}
+            >
+          >
+        : any)
+  >,
+  TRight extends OptionallyHasPreAuthorize<any, any>
+>(left: TLeft, right: TRight): {};
+
+// main preAuthorize HTPipe
+export function HTPipePreAuthorize<
+  TLeft extends OptionallyHasPreAuthorize<any, any>,
+  TRight extends OptionallyHasPreAuthorize<any, any>,
+  TContextInLeft extends TLeft extends HasPreAuthorize<any, any>
+    ? Parameters<TLeft['preAuthorize']>[0]
+    : never,
+  TContextInRight extends TRight extends HasPreAuthorize<any, any>
+    ? Parameters<TRight['preAuthorize']>[0]
+    : never,
+  TContextOutLeft extends TLeft extends HasPreAuthorize<any, any>
+    ? ReturnType<TLeft['preAuthorize']>
+    : never,
+  TContextOutRight extends TRight extends HasPreAuthorize<any, any>
+    ? ReturnType<TRight['preAuthorize']>
+    : never
+>(left: TLeft, right: TRight) {
+  if (isHasPreAuthorize(left) && isHasPreAuthorize(right)) {
+    return {
+      preAuthorize: (
+        context: TContextOutLeft extends TContextInRight
+          ? TContextInLeft
+          : TContextInRight & TContextInLeft
+      ) => {
```

Comment:
This is still branching based on type ... branch based on something meaningful instead, and use early-return to simplify the whole thing.
after getting leftOut, instead...
```
const leftPassed = leftOut===true || (leftOut && (typeof leftOut === "object") && Object.keys(leftOut).length>0);
if(!leftPassed) {
  // early-return a left-fail
  return false;
}
const leftContextOut = (leftOut === true) ? {} : leftOut;
const rightOut = right.preAuthorize(context);
// ...
```
- [ ] clean up

Reply:
in fact...  the `leftPassed = ...` stuff will probably be used 4 times... might wanna factor it off to a authorizationPassed(authOut) function somewhere

Reply:
done